modules/internal/localeModels/flat/LocaleModel.chpl

```diff
@@ -106,19 +106,19 @@ module LocaleModel {
     // The flat memory model assumes only one memory.
     //
     override proc defaultMemory() : locale {
-      return this;
+      return new locale(this);
```

Comment:
(I'm just picking this one as an example, because there are a lot of these in the PR.)

Will this do dynamic memory allocation?  If so, would it be faster to declare a local record with `_instance==this`?  You'd have to check with someone more knowledgeable than I am about the generated code, but I think when we're returning a record like this the caller will already have declared a local one on on the stack and our return value will get copied into that.  Assuming that's so, it seems a shame to go to the work of dynamically allocating memory, filling it in, copying from that, and then freeing it before we return.  If we could just declare a locale record on the stack we'd at least avoid the overhead of the dynamic allocate and free.

Reply:
You are right. 

I can add a `thisWrapper` fields that stores a record that has `this` as its `_instance`.

The reason why I didn't bother doing that is, I was considering whether the whole locale implementation can be refactored a bit in the future, so that these implementation classes return not `locale` but whatever `this.type` is, where the wrapper record can wrap the returned class and return it to the user. Of course, that doesn't probably fix the issue you are bringing up, but it may be a better implementation from a maintainability standpoint. And if we go that way, it may not be easy to achieve what you are suggesting here.

Reply:
This doesn't result in dynamic allocation. The function creates a record in its stack and sets it to be the return argument.